contracts/Registry.sol

```diff
@@ -1,70 +1,604 @@
+
+        requireOrgExistenceConstraint(orgName, true);
+        requireAuthorization(orgName, true);
+        requireServiceExistenceConstraint(orgName, serviceName, false);
+
+        ServiceRegistration memory service;
+        orgsByName[orgName].servicesByName[serviceName] = service;
+        orgsByName[orgName].servicesByName[serviceName].serviceName = serviceName;
+        orgsByName[orgName].servicesByName[serviceName].servicePath = servicePath;
+        orgsByName[orgName].servicesByName[serviceName].agentAddress = agentAddress;
+        orgsByName[orgName].servicesByName[serviceName].orgServiceIndex = orgsByName[orgName].serviceKeys.length;
+        orgsByName[orgName].serviceKeys.push(serviceName);
+
+        for (uint i = 0; i < tags.length; i++) {
+            addTagToServiceRegistration(orgName, serviceName, tags[i]);
+        }
+    }
+
+    function updateServiceRegistration(bytes32 orgName, bytes32 serviceName, bytes32 servicePath, address agentAddress) external {
+
+        requireOrgExistenceConstraint(orgName, true);
+        requireAuthorization(orgName, true);
+        requireServiceExistenceConstraint(orgName, serviceName, true);
+
+        // update the servicePath and agentAddress
+        orgsByName[orgName].servicesByName[serviceName].servicePath = servicePath;
+        orgsByName[orgName].servicesByName[serviceName].agentAddress = agentAddress;
+    }
+
+    function addTagsToServiceRegistration(bytes32 orgName, bytes32 serviceName, bytes32[] tags) external {
+
+        requireOrgExistenceConstraint(orgName, true);
+        requireAuthorization(orgName, true);
+        requireServiceExistenceConstraint(orgName, serviceName, true);
+
+        for (uint i = 0; i < tags.length; i++) {
+            addTagToServiceRegistration(orgName, serviceName, tags[i]);
+        }
+    }
+
+    function addTagToServiceRegistration(bytes32 orgName, bytes32 serviceName, bytes32 tagName) internal {
+
+        // no-op if tag already exists
+        if (orgsByName[orgName].servicesByName[serviceName].tagsByName[tagName].tagName == bytes32(0x0)) {
+            // add the service to the org level tag index
+            Tag memory tagObj;
+            orgsByName[orgName].servicesByName[serviceName].tagsByName[tagName] = tagObj;
+            orgsByName[orgName].servicesByName[serviceName].tagsByName[tagName].tagName = tagName;
+            orgsByName[orgName].servicesByName[serviceName].tagsByName[tagName].itemTagIndex = orgsByName[orgName].servicesByName[serviceName].tags.length;
+            orgsByName[orgName].servicesByName[serviceName].tagsByName[tagName].globalTagIndex = servicesByTag[tagName].orgNames.length;
+            orgsByName[orgName].servicesByName[serviceName].tags.push(tagName);
+
+            // add the service to the global tag index creating a list object for this tag if it does not already exist
+            if (!servicesByTag[tagName].valid) {
+                ServiceOrTypeRepositoryList memory listObj;
```

Comment:
Question about this: in all the cases above (such as creating the organization or service registration structs), we created the struct in memory, immediately put it in the map, and then did all of the field assignments like `mapName[keyName].fieldName = ...`

Is there any difference between that versus doing the assignment in memory and *then* putting the struct in the storage map as we are here?

Reply:
never tested this. going to take this as a work item separate from this PR